## Part 1: Preprocessing

In [52]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras import layers

#  Import and read the attrition data
attrition_df = pd.read_csv('https://static.bc-edx.com/ai/ail-v-1-0/m19/lms/datasets/attrition.csv')
attrition_df.head(3)

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   41       Yes      Travel_Rarely                   Sales                 1   
1   49        No  Travel_Frequently  Research & Development                 8   
2   37       Yes      Travel_Rarely  Research & Development                 2   

   Education EducationField  EnvironmentSatisfaction  HourlyRate  \
0          2  Life Sciences                        2          94   
1          1  Life Sciences                        3          61   
2          2          Other                        4          92   

   JobInvolvement  ...  PerformanceRating RelationshipSatisfaction  \
0               3  ...                  3                        1   
1               2  ...                  4                        4   
2               2  ...                  3                        2   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  \
0                 0                 8                      0               1   
1                 1                10                      3               3   
2                 0                 7                      3               3   

   YearsAtCompany  YearsInCurrentRole  YearsSinceLastPromotion  \
0               6                   4                        0   
1              10                   7                        1   
2               0                   0                        0   

   YearsWithCurrManager  
0                     5  
1                     7  
2                     0  

[3 rows x 27 columns]

In [53]:
# attrition_df.shape

In [54]:
# Determine the number of unique values in each column.
attrition_df.nunique().sort_values(ascending=False).to_frame("Unique Values")

Unique Values
HourlyRate                           71
Age                                  43
TotalWorkingYears                    40
YearsAtCompany                       37
DistanceFromHome                     29
YearsInCurrentRole                   19
YearsWithCurrManager                 18
YearsSinceLastPromotion              16
PercentSalaryHike                    15
NumCompaniesWorked                   10
JobRole                               9
TrainingTimesLastYear                 7
EducationField                        6
JobLevel                              5
Education                             5
EnvironmentSatisfaction               4
JobInvolvement                        4
JobSatisfaction                       4
RelationshipSatisfaction              4
StockOptionLevel                      4
WorkLifeBalance                       4
Department                            3
BusinessTravel                        3
MaritalStatus                         3
OverTime                              2
Attrition                             2
PerformanceRating                     2

In [55]:
#attrition_df.isnull().sum()
# attrition_df.isnull().sum()

In [56]:
## Statistical description of dataset
# round(attrition_df.describe().T,2)

In [57]:
# attrition_df.info()#dtypes

In [58]:
# attrition_df.columns

In [59]:
# attrition_df.describe(include = "object")

## Isolating Object (Non_Numeric) Datatypes

In [60]:
attrition_df.describe(include = "object").info()

<class 'pandas.core.frame.DataFrame'>
Index: 4 entries, count to freq
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Attrition       4 non-null      object
 1   BusinessTravel  4 non-null      object
 2   Department      4 non-null      object
 3   EducationField  4 non-null      object
 4   JobRole         4 non-null      object
 5   MaritalStatus   4 non-null      object
 6   OverTime        4 non-null      object
dtypes: object(7)
memory usage: 256.0+ bytes


## Isolating Numeric Datatypes

In [61]:
attrition_df.describe(include = "number").info()

<class 'pandas.core.frame.DataFrame'>
Index: 8 entries, count to max
Data columns (total 20 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Age                       8 non-null      float64
 1   DistanceFromHome          8 non-null      float64
 2   Education                 8 non-null      float64
 3   EnvironmentSatisfaction   8 non-null      float64
 4   HourlyRate                8 non-null      float64
 5   JobInvolvement            8 non-null      float64
 6   JobLevel                  8 non-null      float64
 7   JobSatisfaction           8 non-null      float64
 8   NumCompaniesWorked        8 non-null      float64
 9   PercentSalaryHike         8 non-null      float64
 10  PerformanceRating         8 non-null      float64
 11  RelationshipSatisfaction  8 non-null      float64
 12  StockOptionLevel          8 non-null      float64
 13  TotalWorkingYears         8 non-null      float64
 14  TrainingTimes

In [62]:
# attrition_df.describe(include = "number").columns

In [63]:
# attrition_df["Department"].value_counts()

### Creating **y_df** with the Attrition and Department columns

In [64]:
# Create y_df with the Attrition and Department columns

X = attrition_df.drop(columns=["Attrition","Department","BusinessTravel","EducationField","JobRole"]) #"OverTime","MaritalStatus",

y_df = attrition_df[["Attrition", "Department"]]


#y_attrition = attrition_df[["Attrition"]]
#y_department = department_df[["Department"]]

## y Doesnt require to drop first or any value due to need for prediction

In [65]:
print(len(X), len(y_df))

1470 1470


In [66]:
y_df.head(3)

Attrition              Department
0       Yes                   Sales
1        No  Research & Development
2       Yes  Research & Development

In [67]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   DistanceFromHome          1470 non-null   int64 
 2   Education                 1470 non-null   int64 
 3   EnvironmentSatisfaction   1470 non-null   int64 
 4   HourlyRate                1470 non-null   int64 
 5   JobInvolvement            1470 non-null   int64 
 6   JobLevel                  1470 non-null   int64 
 7   JobSatisfaction           1470 non-null   int64 
 8   MaritalStatus             1470 non-null   object
 9   NumCompaniesWorked        1470 non-null   int64 
 10  OverTime                  1470 non-null   object
 11  PercentSalaryHike         1470 non-null   int64 
 12  PerformanceRating         1470 non-null   int64 
 13  RelationshipSatisfaction  1470 non-null   int64 
 14  StockOptionLevel        

### Creating List of Columns for the X-data

In [68]:
# Create a list of at least 10 column names to use as X data
# Create X_df using your selected columns
# Show the data types for X_df

X_data = ['Age', 'DistanceFromHome', 'Education', 'EnvironmentSatisfaction',
       'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction',
       'MaritalStatus', 'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'
]
#X_data

In [69]:
# # Create X_df using your selected columns
X_df = attrition_df[X_data]
X_df.head(3)

Age  DistanceFromHome  Education  EnvironmentSatisfaction  HourlyRate  \
0   41                 1          2                        2          94   
1   49                 8          1                        3          61   
2   37                 2          2                        4          92   

   JobInvolvement  JobLevel  JobSatisfaction MaritalStatus  \
0               3         2                4        Single   
1               2         2                2       Married   
2               2         1                3        Single   

   NumCompaniesWorked  ... PerformanceRating  RelationshipSatisfaction  \
0                   8  ...                 3                         1   
1                   1  ...                 4                         4   
2                   6  ...                 3                         2   

   StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                 0                  8                      0   
1                 1                 10                      3   
2                 0                  7                      3   

   WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
0                1               6                   4   
1                3              10                   7   
2                3               0                   0   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     5  
1                        1                     7  
2                        0                     0  

[3 rows x 22 columns]

In [70]:
# Show the data types for X_df

#X_df.info()# head()
X_df.dtypes#()# head()

Age                          int64
DistanceFromHome             int64
Education                    int64
EnvironmentSatisfaction      int64
HourlyRate                   int64
JobInvolvement               int64
JobLevel                     int64
JobSatisfaction              int64
MaritalStatus               object
NumCompaniesWorked           int64
OverTime                    object
PercentSalaryHike            int64
PerformanceRating            int64
RelationshipSatisfaction     int64
StockOptionLevel             int64
TotalWorkingYears            int64
TrainingTimesLastYear        int64
WorkLifeBalance              int64
YearsAtCompany               int64
YearsInCurrentRole           int64
YearsSinceLastPromotion      int64
YearsWithCurrManager         int64
dtype: object

## Using sklearn to **Train Test Split** the Dataset

In [71]:
# Split the data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y_df, random_state=1) #test_size = 0.2,
#X_train, X_test, y_train, y_test = train_test_split(X, y_attrition, y_department, random_state=1) #test_size = 0.2,
#X_train, X_test, y_train_attrition, y_test_attrition, y_train_department, y_test_department = train_test_split(X, y_attrition, y_department, random_state=1) #test_size = 0.2,

## Encoding / PreProcesing  **y_train and y_test Data**

In [72]:
len(y_df)#.head(3)

1470

In [73]:
len(y_train)

1102

In [74]:
len(y_test)

368

In [75]:
## Double checking the values counts for y
print(attrition_df["Attrition"].value_counts())

Attrition
No     1233
Yes     237
Name: count, dtype: int64


In [76]:
print(attrition_df["Department"].value_counts())

Department
Research & Development    961
Sales                     446
Human Resources            63
Name: count, dtype: int64


In [77]:
# Preprocess y. No need to drop any number
# from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.preprocessing import OneHotEncoder

Attrition_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

Attrition_encoder.fit(np.array(y_train['Attrition']).reshape(-1, 1))
y_train_attrition = Attrition_encoder.transform(np.array(y_train['Attrition']).reshape(-1, 1))
y_test_attrition = Attrition_encoder.transform(np.array(y_test['Attrition']).reshape(-1, 1))


In [78]:
len(y_train_attrition)

1102

In [79]:
y_train_attrition

array([[1., 0.],
       [1., 0.],
       [0., 1.],
       ...,
       [1., 0.],
       [1., 0.],
       [1., 0.]])

In [80]:
# Preprocess y
# from sklearn.preprocessing import LabelEncoder, OneHotEncoder
from sklearn.preprocessing import OneHotEncoder

Department_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

Department_encoder.fit(np.array(y_train['Department']).reshape(-1, 1))
y_train_department =Department_encoder.transform(np.array(y_train['Department']).reshape(-1, 1))
y_test_department = Department_encoder.transform(np.array(y_test['Department']).reshape(-1, 1))

In [81]:
y_test_department

array([[0., 1., 0.],
       [0., 0., 1.],
       [0., 1., 0.],
       ...,
       [0., 1., 0.],
       [0., 1., 0.],
       [0., 1., 0.]])

### Encoding / Preprocessing X_train and X_test data

In [82]:
print(attrition_df["OverTime"].value_counts())

OverTime
No     1054
Yes     416
Name: count, dtype: int64


In [83]:
print(attrition_df["MaritalStatus"].value_counts())

MaritalStatus
Married     673
Single      470
Divorced    327
Name: count, dtype: int64


In [84]:
X_train.head(3)
X_train.columns

Index(['Age', 'DistanceFromHome', 'Education', 'EnvironmentSatisfaction',
       'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction',
       'MaritalStatus', 'NumCompaniesWorked', 'OverTime', 'PercentSalaryHike',
       'PerformanceRating', 'RelationshipSatisfaction', 'StockOptionLevel',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion',
       'YearsWithCurrManager'],
      dtype='object')

In [93]:
# # Preprocess y
# # from sklearn.preprocessing import LabelEncoder, OneHotEncoder
# from sklearn.preprocessing import OneHotEncoder

# Overtime_encoder = OneHotEncoder(sparse_output=False, handle_unknown='ignore')

# Overtime_encoder.fit(np.array(X_train['OverTime']).reshape(-1, 1))
# X_train_overtime = Overtime_encoder.transform(np.array(X_train['OverTime']).reshape(-1, 1))
# X_test_overtime = Overtime_encoder.transform(np.array(X_test['OverTime']).reshape(-1, 1))

In [91]:
# Preprocess "Maritalstatus" column (one-hot encoding)
overtime_encoder = OneHotEncoder(sparse_output=False, drop='first')  # Use drop='first' to avoid dummy variable trap

# Fit and transform the training data
overtime_encoded_train = overtime_encoder.fit_transform(X_train[['OverTime']])

# Transform the testing data
overtime_encoded_test = overtime_encoder.transform(X_test[['OverTime']])

# Create DataFrames for the encoded columns
overtime_columns = overtime_encoder.get_feature_names_out(['OverTime'])
df_overtime_encoded_train = pd.DataFrame(overtime_encoded_train, columns=overtime_columns, index=X_train.index)
df_overtime_encoded_test = pd.DataFrame(overtime_encoded_test, columns=overtime_columns, index=X_test.index)

# Concatenate the encoded columns with the original DataFrames
X_train = pd.concat([X_train.drop(columns='OverTime'), df_overtime_encoded_train], axis=1)
X_test = pd.concat([X_test.drop(columns='OverTime'), df_overtime_encoded_test], axis=1)

In [92]:
# Preprocess "Maritalstatus" column (one-hot encoding)
maritalstatus_encoder = OneHotEncoder(sparse_output=False, drop='first')  # Use drop='first' to avoid dummy variable trap

# Fit and transform the training data
maritalstatus_encoded_train = maritalstatus_encoder.fit_transform(X_train[['MaritalStatus']])

# Transform the testing data
maritalstatus_encoded_test = maritalstatus_encoder.transform(X_test[['MaritalStatus']])

# Create DataFrames for the encoded columns
maritalstatus_columns = maritalstatus_encoder.get_feature_names_out(['MaritalStatus'])
df_maritalstatus_encoded_train = pd.DataFrame(maritalstatus_encoded_train, columns=maritalstatus_columns, index=X_train.index)
df_maritalstatus_encoded_test = pd.DataFrame(maritalstatus_encoded_test, columns=maritalstatus_columns, index=X_test.index)

# Concatenate the encoded columns with the original DataFrames
X_train = pd.concat([X_train.drop(columns='MaritalStatus'), df_maritalstatus_encoded_train], axis=1)
X_test = pd.concat([X_test.drop(columns='MaritalStatus'), df_maritalstatus_encoded_test], axis=1)


In [94]:
X_train.columns

Index(['Age', 'DistanceFromHome', 'Education', 'EnvironmentSatisfaction',
       'HourlyRate', 'JobInvolvement', 'JobLevel', 'JobSatisfaction',
       'NumCompaniesWorked', 'PercentSalaryHike', 'PerformanceRating',
       'RelationshipSatisfaction', 'StockOptionLevel', 'TotalWorkingYears',
       'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany',
       'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager',
       'OverTime_Yes', 'MaritalStatus_Married', 'MaritalStatus_Single'],
      dtype='object')

In [95]:
X_train.head(3)

Age  DistanceFromHome  Education  EnvironmentSatisfaction  HourlyRate  \
464   37                 1          3                        2          59   
512   27                 3          4                        1          54   
480   30                12          4                        2          74   

     JobInvolvement  JobLevel  JobSatisfaction  NumCompaniesWorked  \
464               3         3                4                   4   
512               2         1                4                   0   
480               2         1                1                   1   

     PercentSalaryHike  ...  TotalWorkingYears  TrainingTimesLastYear  \
464                 17  ...                 12                      3   
512                 13  ...                  5                      0   
480                 18  ...                  1                      2   

     WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
464                4               6                   5   
512                3               4                   2   
480                4               1                   0   

     YearsSinceLastPromotion  YearsWithCurrManager  OverTime_Yes  \
464                        1                     2           0.0   
512                        1                     1           0.0   
480                        0                     0           0.0   

     MaritalStatus_Married  MaritalStatus_Single  
464                    0.0                   1.0  
512                    0.0                   1.0  
480                    1.0                   0.0  

[3 rows x 23 columns]

## Scaling the Dataset with Standard Scaler

In [96]:
# # Create a StandardScaler
X_scaler = StandardScaler()

# # Fit the StandardScaler to the training data
X_scaler.fit(X_train)

# # Scale the training and testing data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


## Create, Compile, and Train the Model

In [97]:
## Number of Columns in the X_train to be used for the Analysis
X_train.shape[1]

23

In [98]:
# Find the number of columns in the X training data
X_train.shape[1]

# Create the input layer
X_input = layers.Input(shape=(X_train.shape[1],), name= "input_features")

# Create at least two shared layers
X_shared1 = layers.Dense(64, activation='relu')(X_input)
X_shared2 = layers.Dense(128, activation='relu')(X_shared1)

# # Create the output layer
# X_output = layers.Dense(2, activation='softmax')(X_shared2

In [117]:
# Create a branch for Department
# with a hidden layer and an output layer

# Create the output layer
department_hidden = layers.Dense(32, activation='relu', name='department_hidden')(X_shared2)

#department_output= layers.Dense(y_train_department.shape[1], activation='softmax', name='department_output')(department_hidden)
department_output = layers.Dense(y_train_department.shape[1], activation='sigmoid', name='department_output')(department_hidden)


In [118]:
# Create a branch for Attrition
# with a hidden layer and an output layer

# Create the output layer
attrition_hidden = layers.Dense(32, activation='relu', name='attrition_hidden')(X_shared2)
attrition_output = layers.Dense(y_train_attrition.shape[1], activation='sigmoid', name='attrition_output')(attrition_hidden)


In [119]:
# Create the model
model = Model(inputs=X_input, outputs=[attrition_output, department_output])

# Compile the model
model.compile(optimizer='adam',
              loss={'department_output': 'categorical_crossentropy', 'attrition_output': 'binary_crossentropy'},
              metrics={'department_output': 'accuracy', 'attrition_output': 'accuracy'})

# Summarize the model
model.summary()

Model: "functional_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ Layer (type)              ┃ Output Shape           ┃        Param # ┃ Connected to           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━┩
│ input_features            │ (None, 23)             │              0 │ -                      │
│ (InputLayer)              │                        │                │                        │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense (Dense)             │ (None, 64)             │          1,536 │ input_features[0][0]   │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ dense_1 (Dense)           │ (None, 128)            │          8,320 │ dense[0][0]            │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_hidden (Dense)  │ (None, 32)             │          4,128 │ dense_1[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ department_hidden (Dense) │ (None, 32)             │          4,128 │ dense_1[0][0]          │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ attrition_output (Dense)  │ (None, 2)              │             66 │ attrition_hidden[0][0] │
├───────────────────────────┼────────────────────────┼────────────────┼────────────────────────┤
│ department_output (Dense) │ (None, 3)              │             99 │ department_hidden[0][… │
└───────────────────────────┴────────────────────────┴────────────────┴────────────────────────┘

 Total params: 18,277 (71.39 KB)

 Trainable params: 18,277 (71.39 KB)

 Non-trainable params: 0 (0.00 B)

In [120]:
print(len(X), len(y_test_department), len(y_test_attrition))

1470 368 368


In [121]:
print(len(X), len(y_train_department), len(y_train_attrition))

1470 1102 1102


In [133]:
# Train the model
model.fit(X_train, {'department_output': y_train_department, 'attrition_output':y_train_attrition},
    epochs=20, batch_size=32, validation_split=0.2
)

Epoch 1/20
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - attrition_output_accuracy: 1.0000 - department_output_accuracy: 1.0000 - loss: 0.0022 - val_attrition_output_accuracy: 0.8869 - val_department_output_accuracy: 0.7285 - val_loss: 4.1345
Epoch 2/20
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - attrition_output_accuracy: 1.0000 - department_output_accuracy: 1.0000 - loss: 0.0020 - val_attrition_output_accuracy: 0.8869 - val_department_output_accuracy: 0.7195 - val_loss: 4.1466
Epoch 3/20
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - attrition_output_accuracy: 1.0000 - department_output_accuracy: 1.0000 - loss: 0.0019 - val_attrition_output_accuracy: 0.8869 - val_department_output_accuracy: 0.7285 - val_loss: 4.1703
Epoch 4/20
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step - attrition_output_accuracy: 1.0000 - department_output_accuracy: 1.0000 - loss: 0.0020 - val_attrition_output_accuracy: 0.8959 - val_department_output_accuracy: 0.7195 - val_loss: 4.1622
Epoch 5/20
28/28 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/ste

In [134]:
# Evaluate the model with the testing data
test_results = model.evaluate(X_test_scaled, {'department_output': y_test_department, 'attrition_output': y_test_attrition})
test_results

12/12 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - attrition_output_accuracy: 0.5800 - department_output_accuracy: 0.5723 - loss: 8.3831 


[8.6940336227417, 0.5625, 0.570652186870575]

In [135]:
# Print the accuracy for both department and attrition
print(f"Department Accuracy: {test_results[2]*100:.4f} %")
print(f"Attrition Accuracy:  {test_results[2]*100:.4f} %")

Department Accuracy: 57.0652 %
Attrition Accuracy:  57.0652 %


# Summary

In the provided space below, briefly answer the following questions.

1. Is accuracy the best metric to use on this data? Why or why not?

2. What activation functions did you choose for your output layers, and why?

3. Can you name a few ways that this model might be improved?

YOUR ANSWERS HERE

1. this result is frighteningly good at 100% for both attrition and department. However, these accuracies are probably not the best metric for this data because of the large difference between the output accuracy for both attrition and department and evaluation/validation accuracy at 57%.
Best result was obtained with validation split of 0.2.  

2. I tired both sigmoid and softmax, but I settled for sigmoid, since it is caplbe of doing the job with the right formatting.

3. The model worked very well at 100% but I ma very suspicous of th result.
I think model could be improved by getting more data and may be by using diffrerent dense and activation layers.
Having said this, it will hard be beat 100% as long as its stable.